src/consensus/funding.cpp

```diff
@@ -43,15 +43,17 @@ std::set<FundingStreamElement> GetActiveFundingStreamElements(
     const Consensus::Params& params)
 {
     std::set<std::pair<FundingStreamAddress, CAmount>> requiredElements;
-    for (uint32_t idx = Consensus::FIRST_FUNDING_STREAM; idx < Consensus::MAX_FUNDING_STREAMS; idx++) {
-        // The following indexed access is safe as Consensus::MAX_FUNDING_STREAMS is used
-        // in the definition of vFundingStreams.
-        auto fs = params.vFundingStreams[idx];
-        // Funding period is [startHeight, endHeight)
-        if (fs && nHeight >= fs.get().GetStartHeight() && nHeight < fs.get().GetEndHeight()) {
-            requiredElements.insert(std::make_pair(
-                fs.get().RecipientAddress(params, nHeight),
-                FundingStreamInfo[idx].Value(blockSubsidy)));
+    if (params.NetworkUpgradeActive(nHeight, Consensus::UPGRADE_CANOPY)) {
```

Comment:
Can we have an assertion when adding a funding stream, that its start height is greater than or equal to the Canopy activation height? That would ensure that this cannot be an unintended consensus change.

Reply:
There's an interesting chicken-and-egg problem here. When adding a funding stream after Canopy activation, it must not only start after the Canopy activation height, but it must start after the *current* block height, or else it'll cause a catastrophic invalidation of any blocks above its start height. I'll give some thought to whether there's a better way (in terms of our code organization) to directly tie funding streams to network epochs.